### ISSUE TITLE
[css-cascade-5] Do we need an API that lists the cascade layers in a tree scope?

### ISSUE THREAD

**Issue by xiaochengh** [w3c/csswg-drafts#7015]
It might be useful if there is an API that returns all the cascade layers of a tree scope as a sorted list or a tree. For example, DevTools might need this (source: https://twitter.com/kilianvalkhof/status/1484568350504140804?s=20)

However, for the same purpose, Chrome DevTools doesn't need a public web API. Instead, it retrieves information via [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/), which is also a public API.

So while a layer-listing API seems good on its own, I'm not sure about the exact use cases, and therefore not sure about what the exact result format it should use. (I tried to draft something [here](https://gist.github.com/xiaochengh/437cbc4b9a6fc9afdbd97856d9946901))

So I'm opening this issue to collect ideas and see if we should introduce such an API.

@mirisuzanne

**Comment by tabatkins** [comment]
While this definitely seems useful for DevTools, I'm unsure what use-cases exist for public code, besides just "writing a 3rd-party devtools script".

**Comment by Kilian** [comment]
I would hesitate using "just" there, but yes: I'd use this to write ["third party" devtools](https://polypane.app/blog/polypane-8-better-elements-inspector-new-syncing-features-chromium-98-and-more/#cascade-layer-support), which I've currently done by keeping track of all Cascade layer occurrences and building a tree manually. This is less performant than a native API could provide.

From the CSSStyleSheet API you can "easily" deduct which CSSRule gets precedent by calculating and comparing the specificity of selectors combined with their natural order. Cascade Layers makes this no longer possible, yet between two layer declarations there is no way to tell which takes precedent without (re)traversing the entire set of rules available. An API that provides this functionality seems like a natural extension of the existing CSSOM APIs.

**Comment by tabatkins** [comment]
Yeah, I don't doubt this functionality would be *useful* for such cases, but third-party devtools aren't usually a sufficiently large user population to justify adding new features for, particularly if it *can* be worked out otherwise.

We won't *avoid* helping such a population; it just, by itself, doesn't generally justify a feature. That's why I'm asking if there are any further use-cases.
